### PR TITLE
[Core] Hash and comparator not properly defined

### PR DESCRIPTION
**Pull request by loumalouomega**
This give me many warnings on the last version of clang (9). I moved it to the key_hash file, so everyone can use it

### CODE REVIEW

File: kratos/processes/calculate_embedded_nodal_variable_from_skin_process.h
Comment (loumalouomega) [inline kratos/processes/calculate_embedded_nodal_variable_from_skin_process.h:176]: Here a static_cast from std::size_t to bool happened

Comment (rubenzorrilla) [reply kratos/processes/calculate_embedded_nodal_variable_from_skin_process.h:176]: I think that you spotted an error... I guess that everything was working because so far because the second entry of the map is not used.

**Comment by rubenzorrilla** [comment]
Would you mind waiting until #4805 is finally merged to avoid conflicts? I only need someone to approve it.

File: kratos/processes/calculate_embedded_nodal_variable_from_skin_process.h
Comment (loumalouomega) [reply kratos/processes/calculate_embedded_nodal_variable_from_skin_process.h:176]: ¯\_(ツ)_/¯

Comment (loumalouomega) [reply kratos/processes/calculate_embedded_nodal_variable_from_skin_process.h:176]: This is from the pair, it creates an unique hash. Even if you don't use the second argument of the map, then can be replaced by a set/unordered_set, the hash is important. Maybe the number of elements was no so big to conflict on the hash, but for large problems can be an issue for sure in cases where the first art of the pair is equal.

**Comment by loumalouomega** [comment]
> Would you mind waiting until #4805 is finally merged to avoid conflicts? I only need someone to approve it.

Ok, but this will be a very small conflict, I only change 12 lines

File: kratos/includes/key_hash.h
Comment (pooyan-dadvand) [inline kratos/includes/key_hash.h:350]: There is no hash function for pair in std?

Comment (rubenzorrilla) [reply kratos/includes/key_hash.h:350]: I didn't check since this was an `<int,int>` to `bool` map.

File: kratos/processes/calculate_embedded_nodal_variable_from_skin_process.h
Comment (pooyan-dadvand) [reply kratos/processes/calculate_embedded_nodal_variable_from_skin_process.h:176]: AFAIK `std::hash `is specialized for `std::pair`

File: kratos/includes/key_hash.h
Comment (loumalouomega) [reply kratos/includes/key_hash.h:350]: Apparently not

Comment (pooyan-dadvand) [reply kratos/includes/key_hash.h:350]: AFAIK `std::hash` is specialized for `std::pair`

Comment (loumalouomega) [reply kratos/includes/key_hash.h:350]: I will check the change (the original code is not mine)

File: kratos/processes/calculate_embedded_nodal_variable_from_skin_process.h
Comment (loumalouomega) [reply kratos/processes/calculate_embedded_nodal_variable_from_skin_process.h:176]: In C++11 only these ones are specialized:
https://es.cppreference.com/w/cpp/utility/hash

imagen

File: kratos/includes/key_hash.h
Comment (loumalouomega) [reply kratos/includes/key_hash.h:350]: I changed for a general std::pair hash, but std::hash does not work for a std::pair directly, I tried and does not compile

File: kratos/processes/calculate_embedded_nodal_variable_from_skin_process.h
Comment (pooyan-dadvand) [reply kratos/processes/calculate_embedded_nodal_variable_from_skin_process.h:176]: you forgot to read the paragraph after above table:

Note: additional specializations for std::pair and the standard container types, ....

Comment (loumalouomega) [reply kratos/processes/calculate_embedded_nodal_variable_from_skin_process.h:176]: "Note: additional specializations for std::pair and the standard container types, as well as utility functions to compose hashes are available in boost.hash"

It is defined on boost, we try to get rid off of boost

Comment (philbucher) [reply kratos/processes/calculate_embedded_nodal_variable_from_skin_process.h:176]: I would say that if it is header only we could have it copied to Kratos
No need for us to reinvent the wheel
Esp if even the cpp-ref points to it

Comment (loumalouomega) [reply kratos/processes/calculate_embedded_nodal_variable_from_skin_process.h:176]: For me is not a problem

**Comment by rubenzorrilla** [comment]
@loumalouomega just to let you know. @KratosMultiphysics/technical-committee decided to mimic the Boost implementation in here. Yo did already implement the `HashCombine` method as it is in there so it was as simple as computing the seed using such method for both entries of the pair.

I took the chance to remove the `unordered_map` by an `unordered_set` since the key value is actually not needed.

In any case, many thanks for spotting this and for the prior implementation.

**Review by loumalouomega — COMMENTED** [review]
I cannot approve a PR since I created myself

**Comment by rubenzorrilla** [comment]
> I cannot approve a PR since I created myself

No worries, either @philbucher or @pooyan-dadvand can approve it. :wink:

**Comment by philbucher** [comment]
> > I cannot approve a PR since I created myself
> 
> No worries, either @philbucher or @pooyan-dadvand can approve it. wink

I am really not the expert on hashing, please @pooyan-dadvand take the lead here :)

**Comment by philbucher** [comment]
ping @pooyan-dadvand

**Comment by rubenzorrilla** [comment]
@pooyan-dadvand I think this is ready to be merged.

**Review by pooyan-dadvand — APPROVED** [review]
👍

**Comment by loumalouomega** [comment]
Thanks

**Pull request merged**
